### PR TITLE
Ignoring backspace for empty lines on stdin/stdout mode

### LINKED ISSUE #8193: Crash on backspace at start of line, in stdin/stdout mode
Steps to reproduce
------------------
Run the following python code:

```
import subprocess
ps = subprocess.Popen("powershell.exe", stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
ps.stdin.write(b'\b')
ps.stdin.flush()
```

Expected behavior
-----------------
Since no character was sent, the backspace should be ignored

Actual behavior
---------------
Crash, at ConsoleHostUserInterface.cs:1390.
``` sb.Remove(sb.Length - 1, 1); ``` is called, but sb.Length is 0

https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs#L1386-L1396

How to fix
----------------
There should be a check for the length of the string builder, and the code should ignore the backspace if the length is 0

### PR DESCRIPTION
**Pull request by bugale**
## PR Summary

Fix #8193.
On stdin/stdout mode (without console), when a backspace was sent, the code tried to remove the last character in the current line string builder. this was true even when the string builder is empty, thus crashing.
I fixed the issue by ignoring the backspace when the string builder is empty, not echoing it, and not trying to remove from the string builder.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link: 
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs [hunk 1]
```diff
@@ -1365,6 +1365,7 @@ private string ReadLineFromFile(string initialContent)
                 }
 
                 var c = unchecked((char)inC);
+                if (c == '\b' && !NoPrompt && sb.Length == 0) continue;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: I'd expect the check in line below `if (c == '\b' && !NoPrompt && && sb.Length > 0)

**Review by iSazonov — COMMENTED** [review]
Please add test.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs (hunk 1 above)
Comment (bugale) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: What do you mean?
After the Console.Out.Write?
The user of this code have only the prompt at this stage, so echoing him the backspace will make him think the prompt was deleted.

**Comment by iSazonov** [comment]
New tests should be in https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Host/ConsoleHost.Tests.ps1
In the file you can find samples with running PowerShell process with redirections.

**Comment by SteveL-MSFT** [comment]
@bugale for the test, you can do the same as your initial crash report but with PowerShell:

```powershell
$si = [System.Diagnostics.ProcessStartInfo]::new()
$si.FileName = "pwsh"
$si.RedirectStandardInput = $true
$si.RedirectStandardOutput = $true
$si.RedirectStandardError = $true
$pwsh = [System.Diagnostics.Process]::Start($si)
$pwsh.StandardInput.Write("`b")
$pwsh.StandardInput.Write("`b")
```

The second write causes a broken pipe on my macBook.  However, it doesn't fail on my Windows machine.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: It seems that this is really an edge case where there is nothing in the stringbuilder so a backspace in this case is the same as being at the prompt, typing backspace and nothing happens.  In general, I would agree we should not filter the input stream, but this specific case seems ok.

Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: @bugale Please follow our code convention:
```suggestion
                if (c == '\b' && !NoPrompt && sb.Length == 0)
                {
                    continue;
                }
```

Comment (bugale) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: Okay, but exactly one line below that there is a code that uses this conventions

Comment (iSazonov) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: Skipping braces is dangerous. We should use only good patterns in new code. We don't request to fix all old code in the method but you can do this if you want.

Comment (bugale) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs]: Yeah, I know Apple's "goto fail" :) Ill change the other one too

**Comment by bugale** [comment]
I have added a test. I validated it fails before my change, and passes after it.
Unfortunately, it takes 10 seconds to pass, but when the crash happens, it takes ~5 seconds, so I couldn't think of a better way.

**Comment by iSazonov** [comment]
> it takes 10 seconds to pass

Are you sure that it is not timeout?

**Comment by bugale** [comment]
It is a timeout, for the function WaitForExit.
I do something that crashes the shell without my fix, but it takes ~5 seconds to crash.
So in the test, I wait for the crash with a timeout of 10 seconds. If the timeout was reached, it means the process did not crash and there is no bug.
Without my fix, though, the shell crashes before the timeout passes, which fails the test.

**Comment by iSazonov** [comment]
😕 This is not a reliable test - we can get a "race condition". @SteveL-MSFT I believe we should think about xUnit test.

**Comment by bugale** [comment]
> 😕 This is not a reliable test - we can get a "race condition". @SteveL-MSFT I believe we should think about xUnit test.

I changed it now to rely on the output

**Comment by iSazonov** [comment]
PowerShell-CI-linux termporary failed and was restarted.

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Comment by iSazonov** [comment]
@bugale Thanks for your contribution.

**Pull request merged**